paparazzi/src/main/java/app/cash/paparazzi/PaparazziTestExtension.kt

```diff
@@ -0,0 +1,465 @@
+
+    val frameHandler = snapshotHandler.newFrameHandler(snapshot, frameCount, fps)
+    frameHandler.use {
+      val viewGroup = bridgeRenderSession.rootViews[0].viewObject as ViewGroup
+      try {
+        withTime(0L) {
+          // Initialize the choreographer at time=0.
+        }
+
+        viewGroup.addView(view)
+        for (frame in 0 until frameCount) {
+          val nowNanos = (startNanos + (frame * 1_000_000_000.0 / fps)).toLong()
+          withTime(nowNanos) {
+            val result = renderSession.render(true)
+            if (result.status == ERROR_UNKNOWN) {
+              throw result.exception
+            }
+
+            var image = bridgeRenderSession.image
+            renderExtensions.forEach {
+              image = it.render(snapshot = snapshot, view = view, image = image)
+            }
+            frameHandler.handle(scaleImage(image))
+          }
+        }
+      } finally {
+        viewGroup.removeView(view)
+      }
+    }
+  }
+
+  private fun withTime(
+    timeNanos: Long,
+    block: () -> Unit
+  ) {
+    val frameNanos = TIME_OFFSET_NANOS + timeNanos
+
+    // Execute the block at the requested time.
+    System_Delegate.setBootTimeNanos(frameNanos)
+    System_Delegate.setNanosTime(frameNanos)
+    Choreographer.getInstance().doFrame(frameNanos, 0)
+    try {
+      block()
+    } finally {
+      System_Delegate.setNanosTime(0L)
+      System_Delegate.setBootTimeNanos(0L)
+    }
+  }
+
+  private fun createRenderSession(sessionParams: SessionParams): RenderSessionImpl {
+    val renderSession = RenderSessionImpl(sessionParams)
+    renderSession.setElapsedFrameTimeNanos(0L)
+    RenderSessionImpl::class.java
+        .getDeclaredField("mFirstFrameExecuted")
+        .apply {
+          isAccessible = true
+          set(renderSession, true)
+        }
+    return renderSession
+  }
+
+  private fun createBridgeSession(
+    renderSession: RenderSessionImpl,
+    result: Result
+  ): BridgeRenderSession {
+    try {
+      val bridgeSessionClass = Class.forName("com.android.layoutlib.bridge.BridgeRenderSession")
+      val constructor =
+        bridgeSessionClass.getDeclaredConstructor(RenderSessionImpl::class.java, Result::class.java)
+      constructor.isAccessible = true
+      return constructor.newInstance(renderSession, result) as BridgeRenderSession
+    } catch (e: Exception) {
+      throw RuntimeException(e)
+    }
+  }
+
+  private fun scaleImage(image: BufferedImage): BufferedImage {
+    val maxDimension = Math.max(image.width, image.height)
+    val scale = THUMBNAIL_SIZE / maxDimension.toDouble()
+    return ImageUtils.scale(image, scale, scale)
+  }
+
+  private fun ExtensionContext.toTestName(): TestName {
+    val fullQualifiedName = testClass.get().name
+    val packageName = fullQualifiedName.substringBeforeLast('.', missingDelimiterValue = "")
+    val className = fullQualifiedName.substringAfterLast('.')
+    return TestName(packageName, className, testMethod.get().name)
+  }
```

Comment:
This was an extension function on `Description` in the `Paparazzi` class. Here we need to do the same from the JUnit 5 `ExtensionContext`.